File: BST/simple_bst/wisdomgo/cpp-version/src/bst.cpp
#include "bst.h"
#include <stdexcept>
#include <iostream>

BST::BST() {
    root = nullptr;
}

BST::~BST() {
    destroyTree(root);
}

void BST::insert(int val) {
    root = insert(root, val);
}

bool BST::remove(int val) { 
    bool isDeleted = false;
    root = remove(root, val, isDeleted);
    return isDeleted; 
}

bool BST::find(int val) {
    BSTnode* node = root;
    while (node != nullptr) {
        if (val < node->val) {
            node = node->left;
        } else if (val > node->val) {
            node = node->right;
        } else {
            return true;
        }
    }
    return false;
}

int BST::findmin() {
    BSTnode* node = findmin(root);
    if (node == nullptr) {
        throw std::runtime_error("Error: It is an empty tree");
    }
    return node->val;
}

int BST::findmax() {
    BSTnode* node = findmax(root);
    if (node == nullptr) {
        throw std::runtime_error("Error: It is an empty tree");
    }
    return node->val;
}

BSTnode* BST::insert(BSTnode* node, int val) {
    if (node == nullptr) {
        return new BSTnode(val);
    } else if (val < node->val) {
        node->left = insert(node->left, val);
    } else if (val > node->val) {
        node->right = insert(node->right, val);
    }
    return node;
}

BSTnode* BST::remove(BSTnode* node, int val, bool& isDeleted) {
    if (node == nullptr) {
        isDeleted = false;
        return nullptr;
    } else if (val < node->val) {
        node->left = remove(node->left, val, isDeleted);
    } else if (val > node->val) {
        node->right = remove(node->right, val, isDeleted);
    } else {
        isDeleted = true;
        if (node->left == nullptr && node->right == nullptr) {
            delete node;
            return nullptr;
        } else if (node->left == nullptr) {
            BSTnode* temp = node->right;
            delete node;
            return temp;
        } else if (node->right == nullptr) {
            BSTnode* temp = node->left;
            delete node;
            return temp;
        } else {
            BSTnode* temp = findmin(node->right);
            node->val = temp->val;
            node->right = remove(node->right, temp->val, isDeleted);
        }
    }
    return node;
}

BSTnode* BST::findmin(BSTnode* node) {
    if (node == nullptr) {
        return nullptr;
    } else if (node->left == nullptr) {
        return node;
    } else {
        return findmin(node->left);
    }
}

BSTnode* BST::findmax(BSTnode* node) {
    if (node == nullptr) {
        return nullptr;
    } else if (node->right == nullptr) {
        return node;
    } else {
        return findmax(node->right);
    }
}

void BST::destroyTree(BSTnode* node) {
    if (node != nullptr) {
        destroyTree(node->left);
        destroyTree(node->right);
        delete node;
    }
}

File: BST/simple_bst/lingkerio/OCaml-version/simple_bst/bin/bst.ml
(* lib/BST.ml *)

(* Define the type for a binary search tree *)
type 'a bst =
  | Empty
  | Node of 'a * 'a bst * 'a bst

(* Create an empty BST *)
let empty = Empty

(* Insert a value into the BST *)
let rec insert x = function
  | Empty -> Node (x, Empty, Empty)
  | Node (y, left, right) as node ->
    if x < y then Node (y, insert x left, right)
    else if x > y then Node (y, left, insert x right)
    else node

(* Check if a value exists in the BST *)
let rec member x = function
  | Empty -> false
  | Node (y, left, right) ->
    if x < y then member x left
    else if x > y then member x right
    else true

(* Find the minimum value in the BST *)
let rec min_bst = function
  | Empty -> failwith "Tree is empty"
  | Node (x, Empty, _) -> x
  | Node (_, left, _) -> min_bst left

(* Find the maximum value in the BST *)
let rec max_bst = function
  | Empty -> failwith "Tree is empty"
  | Node (x, _, Empty) -> x
  | Node (_, _, right) -> max_bst right

(* Delete a value from the BST *)
let rec delete x = function
  | Empty -> Empty
  | Node (y, left, right) ->
    if x < y then Node (y, delete x left, right)
    else if x > y then Node (y, left, delete x right)
    else
      match left, right with
      | Empty, _ -> right
      | _, Empty -> left
      | _, _ ->
        let min_right = min_bst right in
        Node (min_right, left, delete min_right right)

